src/dataflow/src/render/reduce.rs

```diff
@@ -31,23 +76,427 @@ use super::context::Context;
+/// of holding onto all values.
+#[derive(Clone, Debug)]
+struct MonotonicPlan {
+    /// All of the aggregations we were asked to compute.
+    aggr_funcs: Vec<AggregateFunc>,
+    /// Set of "skips" or calls to `nth()` an iterator needs to do over
+    /// the input to extract the relevant datums.
+    skips: Vec<usize>,
+}
+
+/// Plan for computing a set of hierarchical aggregations
+/// with non-monotonic inputs.
+///
+/// To perform hierarchical aggregations with stable runtimes
+/// under updates we'll subdivide the group key into buckets, compute
+/// the reduction in each of those subdivided buckets and then combine
+/// the results into a coarser bucket (one that represents a larger
+/// fraction of the original input) and redo the reduction in another
+/// layer. Effectively, we'll construct a min / max heap out of a series
+/// of reduce operators (each one is a separate layer).
+#[derive(Clone, Debug)]
+struct BucketedPlan {
+    /// All of the aggregations we were asked to compute.
+    aggr_funcs: Vec<AggregateFunc>,
+    /// Set of "skips" or calls to `nth()` an iterator needs to do over
+    /// the input to extract the relevant datums.
+    skips: Vec<usize>,
+    // The number of buckets in each layer of the reduction tree. Should
+    // be decreasing, and ideally, a power of two so that we can easily
+    // distribute values to buckets with `value.hashed() % bucketes[layer]`.
+    buckets: Vec<u64>,
+}
+
+/// Plan for computing a set of basic aggregations.
+///
+/// There's much less complexity when rendering basic aggregations.
+/// Each aggregation corresponds to one Differential reduce operator.
+/// That's it. However, we still want to present one final arrangement
+/// so basic aggregations present results with the same interface
+/// (one arrangement containing a row with all results) that accumulable
+/// and hierarchical aggregations do. To provide that, we render an
+/// additional reduce operator whenever we have multiple reduce aggregates
+/// to combine and present results in the appropriate order. If we
+/// were only asked to compute a single aggregation, we can skip
+/// that step and return the arragement provided by computing the aggregation
+/// directly.
+#[derive(Clone, Debug)]
+enum BasicPlan {
+    /// Plan for rendering a single basic aggregation. Here, the
+    /// first element denotes the index in the set of inputs
+    /// that we are aggregating over.
+    Single(usize, AggregateExpr),
+    /// Plan for rendering multiple basic aggregations.
+    /// These need to then be collated together in an additional
+    /// reduction. Each element represents the:
+    /// `(index of the set of the input we are aggregating over,
+    ///   the aggregation function)`
+    Multiple(Vec<(usize, AggregateExpr)>),
+}
+
+/// Plan for collating the results of computing multiple aggregation
+/// types.
+///
+/// TODO: could we express this as a delta join
+#[derive(Clone, Debug, Default)]
+struct CollationPlan {
```

Comment:
Very very minor, but: this could have contained a `Vec<ReducePlan>`, which could allow folks to break apart things that we would fuse for them. That might be over-engineering/thinking it, but wanted to float it for future thinking (not now, as I'm sure it has code implications).

Reply:
I didnt do that because I wanted to enforce that `CollationPlan` only has 3 child plans, those child plans are all different types, and none of the child plans are Distinct / CollationPlan with the compiler

Not saying we shouldn't put things in a vec, but i think then we would have to be careful to make sure e.g. CollationPlans didn't recurse